src/snowflake/snowpark/mock/_stage_registry.py

```diff
@@ -0,0 +1,232 @@
+        # glob supports wildcard '?' and '*' searching
+        list_of_files = glob.glob(local_file_name)
+        result_df = TableEmulator(
+            columns=PUT_RESULT_KEYS,
+            sf_types={
+                "source": ColumnType(StringType(), True),
+                "target": ColumnType(StringType(), True),
+                "source_size": ColumnType(DecimalType(10, 0), True),
+                "target_size": ColumnType(DecimalType(10, 0), True),
+                "source_compression": ColumnType(StringType(), True),
+                "target_compression": ColumnType(StringType(), True),
+                "status": ColumnType(StringType(), True),
+                "message": ColumnType(StringType(), True),
+            },
+            dtype=object,
+        )
+
+        if not list_of_files:
+            raise SnowparkSQLException(f"File doesn't exist: {local_file_name}")
+
+        for local_file_name in list_of_files:
+
+            file_name = os.path.basename(local_file_name)
+            stage_target_dir_path = os.path.join(self._working_directory, stage_prefix)
+
+            if not os.path.exists(stage_target_dir_path):
+                os.makedirs(stage_target_dir_path)
+
+            if (
+                os.path.isfile(os.path.join(stage_target_dir_path, file_name))
+                and not overwrite
+            ):
+                status = "SKIPPED"
+            else:
+                shutil.copy(local_file_name, stage_target_dir_path)
+                status = "UPLOADED"
+
+            file_size = os.path.getsize(os.path.expanduser(local_file_name))
+            result_df.loc[len(result_df)] = {
+                k: v
+                for k, v in zip(
+                    PUT_RESULT_KEYS,
+                    [
+                        file_name,
+                        file_name,
+                        file_size,
+                        file_size,
+                        "NONE",
+                        "NONE",
+                        status,
+                        "",
+                    ],
+                )
+            }
+        return result_df
+
+    def upload_stream(
+        self,
+        input_stream: IO[bytes],
+        stage_prefix: str,
+        file_name: str,
+        overwrite: bool = False,
+    ) -> Dict:
+        stage_target_dir_path = os.path.join(self._working_directory, stage_prefix)
+
+        if not os.path.exists(stage_target_dir_path):
+            os.makedirs(stage_target_dir_path)
+
+        status = "UPLOADED"
+        if (
+            os.path.isfile(os.path.join(stage_target_dir_path, file_name))
+            and not overwrite
+        ):
+            status = "SKIPPED"
+        else:
+            with open(os.path.join(stage_target_dir_path, file_name), "wb") as f:
+                try:
+                    f.write(input_stream.read())
+                except ValueError as exc:
+                    raise SnowparkUploadFileException(
```

Comment:
Question: why do we need this?

Reply:
this was originally added to pass the test_put_stream_negative in which case we read a closed stream.
but according to our previous discussion, we choose not to wrap the error.

I have logged a JIRA for error experience before GA: https://snowflakecomputing.atlassian.net/browse/SNOW-1235716.